src/liballoc/collections/btree/search.rs

```diff
@@ -61,16 +61,18 @@ where
 {
     // This function is defined over all borrow types (immutable, mutable, owned),
     // and may be called on the shared root in each case.
-    // Crucially, we use `keys()` here, i.e., we work with immutable data.
-    // `keys_mut()` does not support the shared root, so we cannot use it.
     // Using `keys()` is fine here even if BorrowType is mutable, as all we return
     // is an index -- not a reference.
-    for (i, k) in node.keys().iter().enumerate() {
-        match key.cmp(k.borrow()) {
-            Ordering::Greater => {}
-            Ordering::Equal => return (i, true),
-            Ordering::Less => return (i, false),
+    let len = node.len();
```

Comment:
Just to make sure my reasoning is correct: this *is* safe to call, as any node (including the shared root) has a valid prefix NodeHeader such that we can read the length out, right?

Reply:
(Maybe an easier and more performant way of doing this is to say `!node.is_shared_root()`?

Reply:
I think it might be less performant, because `is_shared_root` needs to fetch the address of the shared root? While we already need to access length anyway. Or does the linker sprinkle that address into every inlined occurrence of `is_shared_root`?

Reply:
And yes,

> this _is_ safe to call, as any node (including the shared root) has a valid prefix NodeHeader

Reply:
As to `into_key_slice`, surely you mean the same for `into_val_slice` and so many more?

I have a hard time understanding when something is deemed safe, but I figured that there is no source of unsafety because the caller needs to pass on ownership of `self`. That the precondition is not asserted or otherwise panicked about, also happens in `push` and `push_front` and there the comment "Necessary for correctness, but this is an internal module" seems to bless doing debug_assert on them. `into_key_slice` isn't even public. Note that `into_val_slice_mut` had its `debug_assert!(!self.is_shared_root());` before I started pushing things around.

Reply:
As to the `search_linear` body, I'm fine with it either way. It's just that I started off with a `is_shared_root` condition and thought I optimized it.

Note that there is another alternative: instead of exposing publicly (in an internal module) that k,v-pairs are stored in the shape of key slices, provide only individual access to keys, [which basically simplifies the search code](https://github.com/rust-lang/rust/commit/58d89467ef83d40340833198c05860492af66cf9#diff-7f15a435b4b02b2a3de6fd1d408a718c). Regardless of whether you also apply that idea on the rest of the callers of `keys`, as in [another upcoming PR](https://github.com/rust-lang/rust/compare/master...ssomers:slice_slasher).

Reply:
This is fine as is, I guess -- not sure as to the len() vs is_shared_root() question performance wise, but ultimately doesn't matter.

I think the rest of the commentary is not currently relevant so not going to reply much to it, but I will briefly mention that the debug_assert is fine because private function pattern isn't really correct at least in my eyes, or rather is too simplistic -- we need to verify that the callers are correct regardless of the debug asserts, which is not always easy (or possible), in which case we should leave it as an assert.